### DISCUSSION TITLE
Why is access to the path `/app/ocelot.json` denied?

### DISCUSSION THREAD

**Discussion by MuhammadAshouri** [ThreeMammals/Ocelot#1956] [Q&A] [answered]
I'm using ocelot in dotnet 8
Running app on windows (self host) is totally fine
But when I run app in docker container I'm getting access denied in AddJsonFile
I've checked container folders and /app/ocelot.json is present (but it's read-only)

Here is my `Program.cs`
```csharp
builder.Configuration
    .SetBasePath(environment.ContentRootPath).AddJsonFile("ocelot.json", false, true)
    .AddOcelot(environment)
    .AddEnvironmentVariables();
builder.Services
    .AddOcelot(configuration);
```
and even when I'm trying to remove read-only attribute I'm getting the same error
```csharp
var file = new FileInfo("ocelot.json")
{
    IsReadOnly = false
};
file.CopyTo(Path.Combine(environment.ContentRootPath, "ocelot.json"), true);
```

I've already checked issues and haven't find the solution except this #1216 which seems like it isn't finished yet

## Specifications

  - Version: 22.0.1
  - Platform: Docker (WSL)
  - Subsystem: Windows

**Comment by raman-m** [comment]
Hi Muhammad!
Welcome to Ocelot world! 🐯 

I can't get it, is it a bug or do you ask the question "How to configure"?
What I see now is misconfiguration issue... Your `Program.cs` code is very short! 

Please upload whole solution to GitHub for review, or attach more artifacts: JSON config, Ocelot Startup C# code, and Ocelot app logs with errors/exceptions.
Configuration feature docs:
- [Configuration](https://ocelot.readthedocs.io/en/latest/features/configuration.html), you have to read all subsections carefully
- [Routing](https://ocelot.readthedocs.io/en/latest/features/routing.html), it helps to write correct route definition JSON

-----
> But when I run app in docker container I'm getting access denied in AddJsonFile

Yeap, setup of Docker environment is out of the scope of Ocelot. But, I agree that we don't have much documentation on how to properly run Ocelot in a Docker environment: [Search for "docker"](https://ocelot.readthedocs.io/en/latest/search.html?q=docker).
Seems, we need to write more documentation for Docker. But for now, you could refer to Microsoft Learn docs looking for [Kestrel & Docker](https://www.bing.com/search?q=microsoft+learn+kestrel+docker) articles.

---
> I've checked container folders and /app/ocelot.json is present (but it's read-only)

This is Linux OS in Docker environment issue (WSL is Linux, aka Ubuntu Linux build). Please refer to Linux or Ubuntu documentation, or read more docs for WSL setup.
If you will still be sure that this Ocelot issue when can't write Ocelot file to Linux OS drive then let us know how we could fix this.

**Reply by MuhammadAshouri** [reply]
I know the config is totally fine because it works perfectly on self-host dotnet in windows (so it cannot be ocelot's bug)
the weirdest thing is that if I mount ocelot.json file in docker compose file it works fine in docker too
(Still I'm pretty sure that ocelot.json is present in container without mounting)
Here is my docker compose (apiGateway section):
```
apiGateway:
        image: ghcr.io/..../apigateway:latest
        restart: always
        container_name: apiGateway
        ports:
            - "9999:443"
        volumes:
            - ./ocelot.json:/app/ocelot.json
```
and I know that mounting such a file is terrible in production, so I asked this question

**Reply by raman-m** [reply]
We are not Docker DevOps team... You need to go to Docker community websites and ask "How to mount file system artifacts to ASP.NET app".

**Comment by raman-m** [comment]
@MuhammadAshouri 
> I've already checked issues and haven't find the solution except this #1216 which seems like it isn't finished yet

Regarding #1216 and the Config in-Memory feature... 🆗 We can prioritize this feature...
Would you like to include this to [the next release](https://github.com/ThreeMammals/Ocelot/milestone/3) to have the feature in NuGet package v24.0 ?

In any case, having this feature doesn't mean all your DevOps problems with Docker and virtual environments will go away... Do you understand that?

**Reply by MuhammadAshouri** [reply]
Yeah I do know that it won't fix it all but still the Access Denied error won't happen at least

**Reply by raman-m** [reply, marked as answer]
OK I added the [Feb'24](https://github.com/ThreeMammals/Ocelot/labels/Feb%2724) label [on Feb 23](https://github.com/ThreeMammals/Ocelot/issues/1216#event-11904924562)
We will try to deliver the #1216 feature in Feb'24 release...

**Reply by raman-m** [reply]
❗The feature has been merged to **develop** in commit https://github.com/ThreeMammals/Ocelot/commit/84bd6e4fddec4dbef8056a7b7e9aca410a8aa1ec on March 15, 2024

@MuhammadAshouri You can compile develop version and attach to your project. Release will be later...
